src/Disks/IO/CachedOnDiskWriteBufferFromFile.cpp

```diff
@@ -68,23 +68,41 @@ bool FileSegmentRangeWriter::write(const char * data, size_t size, size_t offset
 
     if (file_segments.empty() || file_segments.back()->isDownloaded())
     {
-        allocateFileSegment(expected_write_offset, is_persistent);
+        allocateFileSegment(expected_write_offset, segment_kind);
     }
+    else
+    {
+        chassert(file_segments.back()->getCurrentWriteOffset() == expected_write_offset);
 
-    auto & file_segment = file_segments.back();
+        if (expected_write_offset != offset)
+        {
+            throw Exception(
+                ErrorCodes::LOGICAL_ERROR,
+                "Cannot write file segment at offset {}, because current write offset is: {}",
+                offset, expected_write_offset);
+        }
+
+        if (file_segments.back()->range().size() == file_segments.back()->getDownloadedSize())
+        {
+            completeFileSegment(*file_segments.back());
+            allocateFileSegment(expected_write_offset, segment_kind);
+        }
+    }
```

Comment:
it seems you marged with master incorrectly, this code is deleted in master

Reply:
Thanks, I removed this lines